cluster/src/main/java/org/apache/iotdb/cluster/log/manage/serializable/SyncLogDequeSerializer.java

```diff
@@ -0,0 +1,310 @@
+    metaFile = SystemFileFactory.INSTANCE
+        .getFile(systemDir + File.separator + "raftLog" + File.separator + "logMeta");
+    init();
+  }
+
+  // only for test
+  public void setMaxRemovedLogSize(long maxRemovedLogSize) {
+    this.maxRemovedLogSize = maxRemovedLogSize;
+  }
+
+  public Deque<Integer> getLogSizeDeque() {
+    return logSizeDeque;
+  }
+
+  // init output stream
+  private void init() {
+    try {
+      if (!logFile.getParentFile().exists()) {
+        logFile.getParentFile().mkdir();
+        logFile.createNewFile();
+        logFile.createNewFile();
+      }
+
+      logOutputStream = new FileOutputStream(logFile, true);
+      metaOutputStream = new FileOutputStream(metaFile, true);
+    } catch (IOException e) {
+      e.printStackTrace();
+    }
+  }
+
+  @Override
+  public void addLast(Log log, LogManagerMeta meta) {
+    ByteBuffer data = log.serialize();
+    int totalSize = 0;
+    // write into disk
+    try {
+      totalSize += ReadWriteIOUtils.write(data, logOutputStream);
+    } catch (IOException e) {
+      e.printStackTrace();
+    }
+
+    logSizeDeque.addLast(totalSize);
+    serializeMeta(meta);
+  }
+
+  @Override
+  public void removeLast(LogManagerMeta meta) {
+    truncateLogIntern(1);
+    serializeMeta(meta);
+  }
+
+  @Override
+  public void truncateLog(int count, LogManagerMeta meta){
+    truncateLogIntern(count);
+    serializeMeta(meta);
+  }
+
+  private void truncateLogIntern(int count){
+    if(logSizeDeque.size() > count){
+      throw new IllegalArgumentException("truncate log count is bigger than total log count");
+    }
+
+    int size = 0;
+    for (int i = 0; i < count; i++) {
+      size += logSizeDeque.removeLast();
+    }
+    // write into disk
+    try {
+      logOutputStream.getChannel().truncate(logFile.length() - size);
+    } catch (IOException e) {
+      e.printStackTrace();
+    }
+  }
+
+  @Override
+  public void removeFirst(int num) {
+    firstLogPosition += num;
+    for (int i = 0; i < num; i++) {
+      removedLogSize += logSizeDeque.removeFirst();
+    }
+
+    // do actual deletion
+    if (removedLogSize > maxRemovedLogSize) {
+      deleteRemovedLog();
+    }
+
+    // firstLogPosition changed
+    serializeMeta(meta);
+  }
+
+  @Override
+  public List<Log> recoverLog() {
+    if (meta == null) {
+      recoverMeta();
+    }
+
+    if (!logFile.exists()) {
+      return new ArrayList<>();
+    }
+
+    List<Log> result = new ArrayList<>();
+    long count = 0;
+    try {
+      FileInputStream logReader = new FileInputStream(logFile);
+      FileChannel logChannel = logReader.getChannel();
+      while (logChannel.position() < logFile.length()) {
+        // actual log
+        if (count >= firstLogPosition) {
+          Log log = readLog(logReader, false);
+          result.add(log);
+        }
+        // removed log, skip
+        else {
+          readLog(logReader, true);
+        }
+        count++;
+      }
```

Comment:
I think if you can also save the file position of the first log (just another 8 bytes in the meta), it will be much easier to skip the removed logs.

Reply:
Sure~ Fixed